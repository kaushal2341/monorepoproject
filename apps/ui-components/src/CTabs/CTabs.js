import React, {PureComponent} from 'react';
import Proptypes from 'prop-types';
import {withRouter} from 'react-router-dom';
import Tabs from "react-bootstrap/Tabs";
import Tab from "react-bootstrap/Tab";
import {TryCatchHandler} from "../../../commons";

class CTabs extends PureComponent {

    state = {
        activeTabKey: ''
    };


    setActiveTabKey = value => this.setState({activeTabKey: value});

    getCurrentUrlAndSetActiveTab = () => {
        this.setActiveTabKey(this.props.location.pathname);
    };

    handleTabSelect = async path => {
        await this.setActiveTabKey(path);
        this.props.history.push(path);
    };

    componentDidMount() {
        this.getCurrentUrlAndSetActiveTab();
    }

    componentDidUpdate(prevProps, prevState) {
        if (prevProps.history) {
            const newPath = prevProps.history.location.pathname;
            const oldPath = prevProps.location.pathname;
            if (newPath !== oldPath) {
                this.setActiveTabKey(newPath);
            } else {
                return false
            }
        } else {
            return false
        }

    }

    getTabElements = (tab) => {
        let title = <> <i className={tab.icon} aria-hidden="true"></i> {tab.name}</>;

        return (<Tab
            id={"tab_" + tab.id}
            key={"tab_" + tab.id}
            title={title}
            eventKey={tab.eventKey}
        >
            {tab.component}
        </Tab>)
    };

    render() {
        const
            {
                id,
                mountOnEnter,
                transition,
                unmountOnExit,
                variant,
                data
            } = this.props;

        return (
            <Tabs
                id={id}
                activeKey={this.state.activeTabKey}
                mountOnEnter={mountOnEnter}
                transition={transition}
                unmountOnExit={unmountOnExit}
                variant={variant}
                onSelect={path => TryCatchHandler.genericTryCatch(this.handleTabSelect(path))}
            >
                {data.map(tab =>
                    this.getTabElements(tab)
                )}
            </Tabs>);
    }
}

CTabs.defaultProps = {
    location: {pathname: '/'},
    history: {location: {pathname: '/'}}
};

React.propTypes = {
    data: Proptypes.array.isRequired,
    id: Proptypes.string, // HTML id attribute E.g:// id="controlled-tab-example" or "uncontrolled-tab-example"
    // or "noanim-tab-example" or "left-tabs-example"
    mountOnEnter: Proptypes.bool,
    transition: Proptypes.bool,
    unmountOnExit: Proptypes.bool,
    variant: Proptypes.string, // variant = tabs or pills
};

export default withRouter(CTabs);