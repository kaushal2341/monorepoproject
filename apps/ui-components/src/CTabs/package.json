{
  "name": "@cogent/c-tabs",
  "main": "CTabs.js",
  "version": "0.0.0",
  "author": "Sabu Shakya",
  "license": "MIT",
  "private": true
}