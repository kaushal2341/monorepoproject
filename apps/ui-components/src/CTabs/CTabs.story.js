import React from 'react';
import {storiesOf} from "@storybook/react";

import CTabs from './CTabs';
import {HashRouter} from 'react-router-dom';

storiesOf('CTabs', module)
    .add('default CTabs', () => {
        const tabData = [
            {
                id: "1",
                name: "Add",
                icon: "fa fa-plus-circle",
                component: '',
                eventKey: "/generalSetup/adminSetup/add"
            },
            {
                id: "2",
                name: "Manage",
                icon: "fa fa-clipboard",
                component: '',
                eventKey: "/generalSetup/adminSetup/manage"
            },
            {
                id: "3",
                name: "Contact",
                icon: "fa fa-phone",
                component: '',
                eventKey: "/generalSetup/adminSetup/contact"
            }];
        return (
            <HashRouter>
                <CTabs data={tabData}/>
            </HashRouter>
        )
    });