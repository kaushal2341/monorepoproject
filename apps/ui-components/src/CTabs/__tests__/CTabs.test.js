import CTabs from '../CTabs';
import Tabs from "react-bootstrap/Tabs";
import React from "react";

expect.addSnapshotSerializer(enzymeSerializer);

describe('CTabs Tests', () => {
    let wrapper;
    const dataForTab = [
        {
            id: "1",
            name: "Add",
            icon: "fa fa-plus-circle",
            component: "",
            eventKey: "/generalSetup/adminSetup/add",
        },
        {
            id: "2",
            name: "Manage",
            icon: "fa fa-clipboard",
            component: "",
            eventKey: "/generalSetup/adminSetup/manage",
        }
    ];

    describe('Tabs Component Tests', () => {

        beforeEach(() => {
                wrapper = shallow(<CTabs.WrappedComponent data={[]}/>);
            }
        );

        test('if CTabs is defined', () => {
            expect(wrapper.find('CTabs')).toBeDefined();
        });

        test('if Tabs Component is defined', () => {
            expect(wrapper.find('Tabs')).toBeDefined();
        });

        test('if Tabs Component contains all required props', () => {
            let propsRequired = [
                'id',
                'activeKey',
                'mountOnEnter',
                'transition',
                'unmountOnExit',
                'variant',
                'onSelect'
            ];
            let propsAvailable = Object.keys(wrapper.find(Tabs).props());
            propsRequired.forEach((propsReq, i) => (
                expect(propsReq).toContain(propsAvailable[i])
            ));
        });

    });

    describe('State Tests', () => {
        let instance;

        beforeEach(() => {
                wrapper = mount(<CTabs.WrappedComponent data={dataForTab}/>);
                instance = wrapper.instance();
            }
        );

        afterAll(() => {
            wrapper.unmount();
        });

        test('if state is defined and has property `activeTabKey`', () => {
            expect(wrapper.state('activeTabKey')).toBeDefined();
        });

        test('if Tabs Component`s prop `activeKey` is assigned state`s activeTabKey value', () => {
            expect(wrapper.find(Tabs).prop('activeKey')).toEqual(wrapper.state('activeTabKey'));
        });

        test('if state`s property activeTabKey is assigned value on component did mount', () => {
            wrapper.setProps({
                location: {
                    pathname: '/add'
                }
            });
            instance.componentDidMount();
            expect(wrapper.state('activeTabKey')).not.toBe('');
        });

        test('if componentDidUpdate lifecycle will be called and state changed on url change', () => {
            instance.componentDidUpdate({
                location: {
                    pathname: '/'
                },
                history: {
                    location: {
                        pathname: '/manage'
                    }
                }
            });
            wrapper.update();
            expect(wrapper.state('activeTabKey')).toBe('/manage');

        });

        describe('Tab select tests', () => {
            let onSelectMock;
            const pathChanged = "/manage";

            beforeEach(() => {
                    onSelectMock = jest.spyOn(instance, 'handleTabSelect');
                    wrapper.update();
                    wrapper.find('Tabs').prop('onSelect')(pathChanged);
                }
            );

            afterAll(() => {
                wrapper.unmount();
            });

            test('if calls select handler function on tab select', () => {
                expect(onSelectMock).toHaveBeenCalled();
            });

            test('if updates state on tab select', () => {
                expect(wrapper.state('activeTabKey')).toBe(pathChanged);
            });
        });

    });

    describe('Tab Component Test', () => {

        beforeEach(() =>
            wrapper = shallow(<CTabs.WrappedComponent data={dataForTab}/>)
        );

        test('if renders at least one `Tab` component', () => {
            expect(wrapper.find('Tab').length).toBeGreaterThanOrEqual(1);
        });

        test('if tab component contains required prop', () => {
            let propsRequired = [
                'id',
                'title',
                'eventKey'
            ];

            let propsAvailable = Object.keys(wrapper.find('#tab_1').props());
            propsRequired.forEach((propsReq, i) => (
                expect(propsReq).toContain(propsAvailable[i])
            ));
        });

    });

    describe('Snapshot Test', () => {
        wrapper = shallow(<CTabs data={dataForTab}/>);
        test('if renders CTab correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});