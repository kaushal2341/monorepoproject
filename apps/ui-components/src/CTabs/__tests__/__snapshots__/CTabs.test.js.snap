// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CTabs Tests Snapshot Test if renders CTab correctly 1`] = `
<ForwardRef
  activeKey="/"
  onSelect={[Function]}
>
  <Tab
    eventKey="/generalSetup/adminSetup/add"
    id="tab_1"
    key="tab_1"
    title={
      <React.Fragment>
         
        <i
          aria-hidden="true"
          className="fa fa-plus-circle"
        />
         
        Add
      </React.Fragment>
    }
  />
  <Tab
    eventKey="/generalSetup/adminSetup/manage"
    id="tab_2"
    key="tab_2"
    title={
      <React.Fragment>
         
        <i
          aria-hidden="true"
          className="fa fa-clipboard"
        />
         
        Manage
      </React.Fragment>
    }
  />
</ForwardRef>
`;
